Make sure weighted gaps are parallel consistent

lindsayad opened the pull request:
Without the new treatment here, one can imagine that we could potentially
add into an LM residual from multiple processes (well not even potentially,
probably with a high likelihood), and hence we would not be correctly
applying the constraint of min(gap, LM), instead we would be applying
something like min(gap_1, LM) + min(gap_2, LM) where the indices refer to the
process ID.

Refs #16961

lindsayad committed:
Add TIMPI support to MetaPhysicL

lindsayad committed:
Make sure weighted gaps are parallel consistent

Without the new treatment here, one can imagine that we could potentially
add into an LM residual from multiple processes (well not even potentially,
probably with a high likelihood), and hence we would not be correctly
applying the constraint of min(gap, LM), instead we would be applying
something like min(gap_1, LM) + min(gap_2, LM) where the indices refer to the
process ID.

Refs #16961

lindsayad committed:
Skip nodes without LM dofs when zeroing inactive LM dofs

this is relevant for problems in which the Lagrange order of the
primal is different from the Lagrange order of the LM

lindsayad committed:
Make sure weighted velocities are parallel consistent

lindsayad committed:
Parallel consistency test coverage

lindsayad committed:
Cover skipping of inactive nodes without LMs

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/869082/) on dbfd6c7 wanted to post the following:

A change of the following file(s) triggered this check:

  scripts/update_and_rebuild_libmesh.sh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

lindsayad commented:
@milljm @loganharbour what am I doing wrong with my conda libmesh build incrementing?

lindsayad commented:
Lol how is the action-only vel-pdass test failing and the non-action version of it works?!

recuero commented:
`snesmf_reuse_base = false` ? any differences in the creation of the lower-d mesh?

lindsayad committed:
Use parameters that have relationship managers in them

I kind of think that the contact action has been giving us incorrect
relationship managers for mortar since its inception and somehow we have
never triggered it in the test harness until
https://civet.inl.gov/job/870326/ ?!

lindsayad commented:
@recuero see my most recent commit for my suspicion ... we will find out soon!

lindsayad commented:
Your guess would be a good one but it looks to me like we set a segmentation fault even before we print a single residual

lindsayad commented:
@jbadger95 you wanted me to toss something else into this PR right?

jbadger95 commented:
I'm actually not so sure about that, it might be fine as is. Essentially I was thinking that when we compute the normalization factor in `ComputeWeightedGapLMMechanicalContact` that we needed to take the L^2 norm since dual shape functions aren't everywhere positive but if we compute them correctly (using the mortar quadrature on dual shape functions like in [#3023](https://github.com/libMesh/libmesh/pull/3023)) then the normalization is fine. I suppose it could be left alone for now since dual shape functions currently don't work with edge dropping.

lindsayad committed:
Try to limit total number of relationship managers added

If an RM asks to be run on the displaced mesh, then we should respect
that and only add it for the displaced mesh. Because in theory that RM
may only function properly if it's really acting with the displaced mesh
(or the converse may be true: a RM may only function properly if it's
acting on the reference mesh). Our proxy geometric relationship managers
should handle ghosting for whatever mesh the RM *doesn't* get added to.
So a summary of what this commit does:

- The same way we only add algebraic ghosting functors to the
  requested undisplaced/displaced dof-map, we now only add geometric ghosting
  functors the requested undisplaced/displaced mesh
- We allow geoemetric proxy relationship managers (which we did not add
  before) to ensure that ghosting is consistent between reference and
  displaced meshes, whereas before we just relied on the idea that cloning
  geometric ghosting functors and hoping they gave the same ghosting would
  work out for us. Thus far in our test suite it had, but who knows what
  would have happened in the future
- With the new existence of geometric proxy relationship managers, we
  needed to make some changes in the `Adaptivity` class in order to
  ensure that the meshes stay roughly in sync such that the
  `ProxyRelationshipManager` can function correctly (e.g. it relies on
  unique id queries so elements cannot be deleted or renumbered before
  both meshes have been adapted)

recuero commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -107,20 +113,59 @@ ComputeFrictionalForceLMMechanicalContact::residualSetup()
+ComputeFrictionalForceLMMechanicalContact::communicateVelocities()
 {
-  residualSetup();
+#ifdef MOOSE_SPARSE_AD
```
Would it make sense to make this class to require sparse patterns?

recuero replied:
Just wondering if we are going to be moving forward with sparse, global framework for mortar.

lindsayad replied:
yes we emit the run-time errors if someone tries to use these mortar objects without the global indexing configuration, but this is required so that we can even compile this code when not using global indexing

recuero commented on conda/libmesh/build.sh:
```diff
@@ -64,6 +64,7 @@ EOF`
 
 ../configure ${BUILD_CONFIG} \
                      --prefix=${PREFIX}/libmesh \
+                     --with-future-timpi-dir=${PREFIX}/libmesh \
```
First use of timpi in MOOSE?

lindsayad replied:
No we use TIMPI all the time, but this will be the first time we enable the parallel communication capabilities of TIMPI with MetaPhysicL types (e.g. `ADReals`)

recuero reviewed (commented):
This is great 👍

recuero reviewed (commented):
I encountered the github unicorn and messed up the review.

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -107,20 +113,59 @@ ComputeFrictionalForceLMMechanicalContact::residualSetup()
+          _nodal ? static_cast<const DofObject *>(lm_mesh.node_ptr(dof_id))
+                 : static_cast<const DofObject *>(lm_mesh.elem_ptr(dof_id));
+      mooseAssert(dof_object, "This should be non-null");
+      _dof_to_weighted_tangential_velocity[dof_object] = std::move(pr.second);
```
Oh wow this is a great catch. That could have lived as a bug for quite a while!

lindsayad commented:
@fdkong I'd like your review on the last commit in this PR

lindsayad commented:
Maybe @roystgnr as well

roystgnr commented on framework/src/base/Adaptivity.C:
```diff
@@ -221,21 +222,88 @@ Adaptivity::adaptMesh(std::string marker_name /*=std::string()*/)
+    // If we have both reference and displaced meshes, then we have to be very careful to keep them
+    // in sync. We have proxy relationship managers that rely on matching ghosted elements via
+    // unique id, so given that, we must make sure that we do not try to delete elements or renumber
+    // them (changing their unique id) at the stage that only one mesh has been refined. We must
```
Wait, what had been going on here?  Mesh renumbering should never change unique_id() on an already-existing object, only id().

lindsayad replied:
I got scared when I saw this code in `DistributedMesh::renumber_dof_objects` and saw `set_unique_id` but I admittedly did not look closely at the surrounding code
```c++
#ifdef LIBMESH_ENABLE_UNIQUE_ID
  auto unique_gather_functor =
    [
#ifndef NDEBUG
     this,
#endif
     &objects]
    (processor_id_type, const std::vector<dof_id_type> & ids,
     std::vector<unique_id_type> & data)
    {
      std::size_t ids_size = ids.size();
      data.resize(ids_size);

      for (std::size_t i=0; i != ids_size; ++i)
        {
          T * obj = objects[ids[i]];
          libmesh_assert(obj);
          libmesh_assert_equal_to (obj->processor_id(), this->processor_id());
          data[i] = obj->valid_unique_id() ? obj->unique_id() : DofObject::invalid_unique_id;
        }
    };

  auto unique_action_functor =
    [&objects]
    (processor_id_type libmesh_dbg_var(pid),
     const std::vector<dof_id_type> & ids,
     const std::vector<unique_id_type> & data)
    {
      for (auto i : index_range(ids))
        {
          T * obj = objects[ids[i]];
          libmesh_assert (obj);
          libmesh_assert_equal_to (obj->processor_id(), pid);
          if (!obj->valid_unique_id() && data[i] != DofObject::invalid_unique_id)
            obj->set_unique_id(data[i]);
        }
    };

  const unique_id_type * unique_ex = nullptr;
  Parallel::pull_parallel_vector_data
    (this->comm(), requested_ids, unique_gather_functor,
     unique_action_functor, unique_ex);
#endif
```
So historically we have not added these `ProxyRelationshipManagers` to the mesh (e.g. as geometric ghosting functors), but now I want to. After adding them to the mesh, I needed to at a minimum prevent remote element deletion when the reference mesh has been refined and the displaced mesh has not. Perhaps disabling renumbering is not required if we're not modifying existing unique ids 🤔

lindsayad replied:
Ah you don't have to look very closely at all do you? Lol
```c++
         if (!obj->valid_unique_id() && data[i] != DofObject::invalid_unique_id)
            obj->set_unique_id(data[i]);
```

roystgnr replied:
That's some old code there ... I think it might be redundant now, actually - we ought to currently have every ghost's unique_id in sync earlier in the process of modifying a mesh.  Once you've got some good test coverage for the kinds of tricky messes fixed by this PR, let me know.  I may want to try out a libMesh PR replacing that "update any object that doesn't already have a valid unique_id" with "scream and die if any object doesn't already have a valid unique_id", and just see what happens.

lindsayad replied:
prod-to-resolve-button @roystgnr :smile:

roystgnr replied:
Ha!  Sorry, yeah, this PR is good to go.  I just need to remember to make that PR in libMesh, see if I'm right about the code there having been made redundant or if there's some special case I'm forgetting.

fdkong commented on framework/src/relationshipmanagers/ProxyRelationshipManager.C:
```diff
@@ -77,8 +77,10 @@ ProxyRelationshipManager::operator()(const MeshBase::const_element_iterator & /*
+  // We have to use all elements because the ghosted elements from the other mesh may very well
+  // include inactive ancestors and things of that sort
+  for (auto elem_it = _moose_mesh->getMesh().elements_begin();
+       elem_it != _moose_mesh->getMesh().elements_end();
```
For distributed mesh, you will get local+ghost, but for replicated mesh, you will have the whole world.  But I guess you only need local + ghost

lindsayad replied:
The thing is that here we are determining ghosting for the current mesh based on ghosting from the other mesh. Local + ghost for the current mesh may very well miss elements that the other mesh is going to tell us to ghost

fdkong replied:
Oh, I see your point. Should we just have a clone somewhere so that one mesh is set and the other will be simple clone of the set one? It might be too late to clone ...

lindsayad replied:
Yes the original reason I had to make this change was because of adaptivity between time-steps which happens way after displaced mesh creation (where we do indeed just clone the reference mesh)

lindsayad replied:
It's back to https://github.com/libMesh/libmesh/issues/2661

fdkong commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -107,20 +113,59 @@ ComputeFrictionalForceLMMechanicalContact::residualSetup()
+
+  auto action_functor = [this, &lm_mesh](const processor_id_type libmesh_dbg_var(pid),
+                                         const std::vector<Datum> & sent_data) {
+    mooseAssert(pid != this->processor_id(), "We do not send messages to ourself here");
```
Send message to yourself is cheap, actually

lindsayad replied:
It may be cheap but the way I've setup the logic is that on the current process I don't setup pushes to myself so I want to make sure that I'm actually doing what I think I'm doing

fdkong reviewed (changes requested):
It looks good

lindsayad commented on framework/src/base/Adaptivity.C:
```diff
@@ -221,21 +222,88 @@ Adaptivity::adaptMesh(std::string marker_name /*=std::string()*/)
+        //
+        // This code comes from the latter half of MeshBase::prepare_for_use
+        //
+        mesh.get_boundary_info().regenerate_id_sets();
```
Perhaps I should copy over the comments from `MeshBase::prepare_for_use`
```c++
  // If we're using DistributedMesh, we'll probably want it
  // parallelized.
  if (this->_allow_remote_element_removal)
    this->delete_remote_elements();

  // Much of our boundary info may have been for now-remote parts of the mesh,
  // in which case we don't want to keep local copies of data meant to be
  // local. On the other hand we may have deleted, or the user may have added in
  // a distributed fashion, boundary data that is meant to be global. So we
  // handle both of those scenarios here
  this->get_boundary_info().regenerate_id_sets();

  if (!_skip_renumber_nodes_and_elements)
    this->renumber_nodes_and_elements();
```

fdkong replied:
Oh, I see. We have a "global" boundary that is an evil case but we have there anyway. 

I wish we could distribute everything. However it is not the issue of PR

lindsayad committed:
Allow renumbering to happen unhindered during adaptivity

cticenhour commented on scripts/update_and_rebuild_libmesh.sh:
```diff
@@ -164,6 +164,7 @@ if [ -z "$go_fast" ]; then
 
   source $SCRIPT_DIR/configure_libmesh.sh
   SRC_DIR=${SCRIPT_DIR}/../libmesh configure_libmesh --prefix=$LIBMESH_DIR \
+                                                     --with-future-timpi-dir=$LIBMESH_DIR \
```
```suggestion
```
Place in `configure_libmesh.sh`

lindsayad replied:
There is no `LIBMESH_DIR` variable in `configure_libmesh.sh`

cticenhour replied:
See my other comment for the equivalent expression.

lindsayad replied:
Basically this configure option should mirror `--prefix`, so if `--prefix` can't be moved to `configure_libmesh.sh`, then neither can `--with-future-timpi-dir`

cticenhour replied:
Though I see how we're missing generalizing here....166 should be changed to use LIBMESH_DIR instead of SCRIPT_DIR. I will adjust my comment.

cticenhour replied:
I see what your intent is here now. Withdrawing my comment, as this is related to where the installed dir will be, not the currently configured one.

cticenhour commented on scripts/configure_libmesh.sh:
```diff
@@ -47,6 +47,7 @@ function configure_libmesh()
                --enable-petsc-hypre-required \
                --enable-metaphysicl-required \
                --with-cxx-std-min=2014 \
+               --without-gdb-command \
```
```suggestion
               --without-gdb-command \
               --with-future-timpi-dir=${SRC_DIR} \
```

SRC_DIR is always the `libmesh` root directory, regardless of which version of configure we're doing.

lindsayad replied:
Like the root source directory? If so, that's not what I want. It needs to be the install directory with `bin,include,lib`

cticenhour replied:
You mean the contents of `build`? Then that would be `${SRC_DIR}/build`.

cticenhour replied:
Nevermind my deleted comment. I understand now. Need more coffee.

cticenhour commented on conda/libmesh/build.sh:
```diff
@@ -75,9 +75,9 @@ fi
 source $SRC_DIR/configure_libmesh.sh
 export INSTALL_BINARY="${SRC_DIR}/build-aux/install-sh -C"
 METHODS="opt oprof devel dbg" configure_libmesh --prefix=${PREFIX}/libmesh \
+                                                --with-future-timpi-dir=${PREFIX}/libmesh \
```
```suggestion
```
Likewise, move to `configure_libmesh.sh`

cticenhour replied:
Withdrawing my comment. I see your intent with "future" TIMPI dir now.

loganharbour replied:
I guess I don't follow this completely

lindsayad replied:
Maybe https://github.com/roystgnr/MetaPhysicL/pull/58 can help explain. We have to make sure that METAPHYSICL_HAVE_TIMPI is defined in metaphysicl_config.h otherwise parallel metaphysicl code will not be available

lindsayad replied:
This PR relies on the ability to send and receive dual numbers

cticenhour reviewed (changes requested):
Limiting my comments to the conda and libmesh configure stuff.

cticenhour reviewed (approved):
Approving conda and libmesh configuration script changes only.

lindsayad commented:
@roystgnr @fdkong are you good with this?
